### PR TITLE
Apply ARM fix to devtools-archlinuxcn

### PR DESCRIPTION
**Pull request by yuyichao**
The new patch added (`0004-*`) fixes the handling of server URL on arm without breaking it on x86 (AFAICT....) The patches applied are tracked at https://github.com/yuyichao/devtools/tree/archlinuxcn right now.

With this, on ARM, one should just install devtools-archlinuxcn to get a working base devtools. devtools-arm-git should work for providing a working arm wrapper (`extra-*-build`) on top of it, and `devtools-cn-git` should provide all the `archlinuxcn-*` wrappers.

### CODE REVIEW

**Comment by yan12125** [comment]
Thank you very much for this. I'm also working on updating devtools-archlinuxcn. Some discussions are needed, especially for the new BUILDTOOL variable. Could you wait me for a few more days before merging this? I need to collect related information.

Some other related discussions: https://github.com/archlinuxcn/repo/pull/2789#issuecomment-1140229167

File: archlinuxcn/devtools-archlinuxcn/0004-Handle-ArchLinuxARM-server-URL-in-arch-nspawn.patch [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ 
+ # shellcheck disable=2016
+-host_mirrors=($(pacman-conf --repo extra Server 2> /dev/null | sed -r 's#(.*/)extra/os/.*#\1$repo/os/$arch#'))
++pacman_arch=$(pacman-conf Architecture)
++if [[ $pacman_arch = arm ]] || [[ $pacman_arch = armv6h ]] || [[ $pacman_arch = armv7h ]] || [[ $pacman_arch = aarch64 ]]; then
++	is_arm=1
++else
++	is_arm=0
++fi
++if ((is_arm)); then
```
Comment (CoelacanthusHex) [inline archlinuxcn/devtools-archlinuxcn/0004-Handle-ArchLinuxARM-server-URL-in-arch-nspawn.patch:25]: I think it is better not to use intermediate variables for such a conditional judgment that has a clear meaning and is only used once?

Comment (yuyichao) [reply archlinuxcn/devtools-archlinuxcn/0004-Handle-ArchLinuxARM-server-URL-in-arch-nspawn.patch:25]: That's fair. An earlier version had this used elsewhere before I decided that it's easier to just set all the variables here....

**Comment by yuyichao** [comment]
> Could you wait me for a few more days before merging this?

Sure. That's why this was submitted as a PR.

Personally, I mainly care about the arm patch and https://github.com/archlinux/devtools/commit/30ed6920c7e7962b7161bc2405c63e9a1ab95be2 .

**Comment by yan12125** [comment]
Here is an issue needing discussions: #2798. It's quite long and not directly related to ARM, so I created a new issue.

**Comment by CoelacanthusHex** [comment]
unshare pacstrap patch has been merged upstream.
We can remove it in the near future.
https://gitlab.archlinux.org/archlinux/devtools/-/merge_requests/90

**Comment by yan12125** [comment]
Upstream devtools 20220609 is out :tada:

**Comment by yuyichao** [comment]
Updated to be based on 20220609 and I've added patch to implement a overwritable BUILDTOOL and also tweaked to make sure we can track upstream BUILDTOOLVER .

**Comment by yan12125** [comment]
Thanks for the update, nice `$BUILDTOOLVER` trick! I have some other minor changes. Do you mind if I push to `devtools-fixarm` branch directly?

**Comment by yuyichao** [comment]
Please do.

**Comment by yan12125** [comment]
Done! I keep additional changes as a separate commit (f0e9666e1e30ab983e643d305ed7085b08a4b8d8). It's better to squash it before merge.

**Comment by yan12125** [comment]
Hmm, just noticed upstream is also discussing how to allow overriding `$BUILDTOOL`: https://gitlab.archlinux.org/archlinux/devtools/-/issues/76. The proposed patch there does not fully match our needs, though.

**Comment by yan12125** [comment]
I think this can be merged?

Just to clarify: I'm not going to rebase 0004-Change-default-BUILDTOOL-to-devtools-archlinuxcn.patch on top of https://gitlab.archlinux.org/archlinux/devtools/-/issues/76 as the latter does not fit our needs.

**Comment by yuyichao** [comment]
The current version still looks good to me. I guess if no one else has any comment I can merge sometime soon.

**Comment by yuyichao** [comment]
> I guess sudoers on build.archlinuxcn.org should be updated if this pull request is merged as-is.

And I assume that's only necessary to allow someone to change BUILDTOOL.

**Comment by yuyichao** [comment]
I was just doing a local test before merging and noticed that the checksums were out of data. Should be fixed now.

**Comment by yuyichao** [comment]
It'll also be good if an admin could create a fork of devtools and track the patches there.

File: archlinuxcn/devtools-archlinuxcn/lilac.yaml [hunk 2]
```diff
@@ -3,4 +3,10 @@
 maintainers:
   - github: farseerfc
 
-managed: false
+post_build: git_pkgbuild_commit
+
+update_on:
+  - source: manual
+    manual: 20220207
```
Comment (yuyichao) [inline archlinuxcn/devtools-archlinuxcn/lilac.yaml:10]: Forgot to update this..... I'll leave it to the next person updating this package to worry about this in a few months....

**Comment by yan12125** [comment]
> And I assume that's only necessary to allow someone to change BUILDTOOL.

Yep

> noticed that the checksums were out of data

Sorry, I forgot to run updpkgsums again after modifying patches.

> It'll also be good if an admin could create a fork of devtools and track the patches there.

Agreed. I don't have the permission to fork https://github.com/archlinux/devtools as https://github.com/archlinuxcn/devtools, either.

**Comment by lilydjwg** [comment]
> > It'll also be good if an admin could create a fork of devtools and track the patches there.
> 
> Agreed. I don't have the permission to fork https://github.com/archlinux/devtools as https://github.com/archlinuxcn/devtools, either.

Forked and added permissions to you.

-- 
Best regards,
lilydjwg

**Comment by yan12125** [comment]
Great, thanks! I pushed current patches to https://github.com/archlinuxcn/devtools/commits/master. Next time devtools-archlinuxcn can switch to that fork. Feel free to update it if you prefer a separate branch for patches.

**Comment by lilydjwg** [comment]
> And I assume that's only necessary to allow someone to change BUILDTOOL.

It's needed anyway. I've updated the sudoers file on build machines.

**Pull request merged**